### PR TITLE
Search locally for git sources before fetching them in online mode

### LINKED ISSUE #2365: Search locally for git sources before fetching them in online mode
While working on #1787, I realized that the implementation for it would also open up the way for some easy speed-ups for online mode as well. Currently we are directly trying to fetch git repos from remote which in the case of `std` takes some time as it fetches the `sway` repo. What we can do (as a TODO comment in the `fetch_git()` suggests) is basically looking for the repo locally first. If we cannot find it we should fallback to fetching the remote repo. 

> I implemented the logic for git sources other than `branch` as it seemed to be easier to start with and I am providing some benchmarks so that we can confirm the searching for a local version is manageable. For comparison i created an empty project which (implicitly) depends on `std`. 
> 
> Building the empty project without searching for a local repo (and doing the fetch & pin stuff takes) **11.5 seconds** on my system 
> Building the empty project with searching for a local repo takes **1.35 seconds** on my system which feels like butter :butter: 
> 
> I think it is worth searching for a local copy even if we are in `online` mode and if we cannot find one then we can fetch & pin (as a `TODO` suggests at the relevant part of the code).

_Originally posted by @kayagokalp in https://github.com/FuelLabs/sway/issues/1787#issuecomment-1192562698_

### PR DESCRIPTION
**Pull request by kayagokalp**
closes #2365.

Searching locally for checkouts is faster than fetching whole repo (at least it is the case for sway repo for std dependency). 
So if the user already fetched a checkout looking for it locally first speeds-up the overall building process but if the git reference is to a branch or default branch, the remote may have a newer commit than the one we have locally. So with this PR, we are first checking if the git reference is to a specific commit (i.e: a tag or rev reference) if that is the case before fetching from remote, we are looking locally. After searching locally if we couldn't find the exact local checkout or an error happened we fallback to fetching from remote. For non specific git references (such as branch or default branch) we are directly fetching from remote to get the latest.

#### My little benchmark (project implicitly depending on std, latest tag already fetched before, lock file does not exists):

**Before this PR:** 3.45 seconds
**After this PR:** 1.82 seconds

## TODO

- [x] Merge #2366

### CODE REVIEW

**Review by JoshuaBatty — APPROVED** [review]
Nice one, utACK.

File: forc-pkg/src/pkg.rs [hunk 1]
```diff
@@ -1338,11 +1338,39 @@ fn pin_pkg(
...
+                    Ok(Some((local_path, commit_hash))) => {
+                        let pinned_git = SourceGitPinned {
+                            source: git_source.clone(),
+                            commit_hash,
+                        };
+                        (pinned_git, local_path)
+                    }
+                    _ => {
+                        // If the checkout we are looking for does not exists locally or an
+                        // error happened during the search fetch it
```
Comment (bingcicle) [inline forc-pkg/src/pkg.rs:1364]:
```suggestion
                        // If the checkout we are looking for does not exist locally or an
                        // error happened during the search, fetch it
```

**Review by bingcicle — APPROVED** [review]
lgtm, some minor comment nits

File: forc-pkg/src/pkg.rs (hunk 1 above)
Comment (kayagokalp) [reply forc-pkg/src/pkg.rs:1364]: Thanks for catching it! Will fix this in the next PR 👍

**Pull request merged**